text/0066-separate-context-propagation.md

```diff
@@ -0,0 +1,559 @@
+its own namespaced set of keys in the context, containing all of the data for 
+that cross-cutting concern.
+
+**SetValue(context, key, value) -> context**  
+To record the local state of an aspect, the Context API provides a function 
+which takes a context, a key, and a value as input, and returns an updated 
+context which contains the new value.
+
+**GetValue(context, key) -> value**  
+To access the local state of an aspect, the Context API provides a function 
+which takes a context and a key as input, and returns a value.
+
+### Optional: Automated Context Management
+When possible, the OpenTelemetry context should automatically be associated 
+with the program execution context. Note that some languages do not provide any 
+facility for setting and getting a current context. In these cases, the user is 
+responsible for managing the current context.  
+
+**SetCurrent(context)**  
+To associate a context with program execution, the Context API provides a 
+function which takes a Context.
+
+**GetCurrent() -> context**  
+To access the context associated with program execution, the Context API 
+provides a function which takes no arguments and returns a Context.
+
+
+## Propagation API
+
+Aspects send their state to downstream processes via propagators: 
+functions which read and write context into RPC requests. Each aspect creates a 
+set of propagators for every type of supported medium - currently only HTTP 
+requests.
+
+**HTTPExtract(context, request) -> context**  
+To receive data injected by prior upstream processes, the Propagation API 
+provides a function which takes a context and an HTTP request, and returns 
+context which represents the state of the upstream system.
+
+**HTTPInject(context, request)**  
+To send the data for all aspects downstream to the next process, the 
+Propagation API provides a function which takes a context and an HTTP request, 
+and mutates the HTTP request to include an HTTP Header representation of the 
+context.
+
+**ChainHTTPInjector(injector, injector) -> injector**  
+To allow multiple aspects to inject their context into the same request, the 
+Propagation API provides a function which takes two injectors, and returns a 
+single injector which calls the two original injectors in order.
+
+**ChainHTTPExtractor(extractor, extractor) -> extractor**  
+To allow multiple aspects to extract their context from the same request, the 
+Propagation API provides a function which takes two extractors, and returns a 
+single extractor which calls the two original extractors in order.
+
+### Optional: Global Propagators
+It is often convenient to create a chain of propagators during program 
+initialization, and then access these combined propagators later in the program. 
+To facilitate this, global injectors and extractors are optionally available. 
+However, there is no requirement to use this feature.
+
+**SetHTTPExtractor(extractor)**  
```

Comment:
Hey @yurishkuro I implemented a very simple approach for Java - let me know what you think about https://github.com/open-telemetry/opentelemetry-java/pull/655#issuecomment-557580901 and also, about whether using the builder pattern would be fine with you (in order to remove the chained calls), i.e.

```
Propagators props = Propagators.newBuilder()
  .addHttpExtractor(TraceContextExtractor())
  .addHttpExtractor(DefaultCorrelationContextExtractor())
  .addHttpExtractor(DefaultBaggageExtractor())
  .build();
OpenTelemetry.setPropagators(props);
...
OpenTelemetry.getPropagators().inject(Context.current(), carrier, carrierSetter);
```

Reply:
@yurishkuro these functions are meant to imply a global or a registry, where you set and get the Propagator. And "Chaining Propagators" is the equivalent to a builder or framework. In the simplest model, there is no need for an extra framework interface - a combined propagator is still just a propagator from the outside.

Rewriting @carlosalberto's Java in pseudocode from this OTEP, it would look like the following:
```
extractor = propagation.ChainHTTPExtractor(
              TraceContextExtractor(), 
              DefaultCorrelationContextExtractor(),
              DefaultBaggageExtractor())

propagation.SetHTTPExtractor(extractor);

...

extractor = propagation.GetHTTPExtractor()
extractor(context.GetCurrent(), headers);
```

Is there something missing from the above example? It does seem like there is confusion about how the globals and chaining are related, maybe there are better words for these functions. But I would like to know if the pseudocode is functionally correct.

Reply:
I'm +1 for a propagator object, and have a global of that as the way to have a global registry. 

The rationalle for named propagators isn't clear to me, but I support the code that @jmacd has written.